sdk/python/feast/infra/online_stores/contrib/mysql_online_store/util.py

```diff
@@ -0,0 +1,325 @@
+
+def default_version_id(feature_view_name: str, offset: int = 0) -> int:
+    offset_bstr = (feature_view_name + '_' + str(offset)).encode('utf-8')
+    return int.from_bytes(hashlib.sha256(offset_bstr).digest()[:8], 'little')
+
+
+def build_legacy_features_read_query(
+    project: str,
+    table: FeatureView,
+    union_offset: Optional[int] = None
+) -> str:
+    base = "SELECT feature_name, value, event_ts" + " " if union_offset is None else f", {union_offset} as __i__ "
+    base += f"FROM {_table_id(project, table)} WHERE entity_key = %s"
+    return base
+
+
+def build_feature_view_features_read_query(
+    config: MySQLOnlineStoreConfig,
+    feature_view_name: str,
+    union_offset: Optional[int] = None
+) -> str:
+    version_table, data_table = config.feature_version_table, config.feature_version_table
+    base = "SELECT feature_name, value, event_ts" + "" if union_offset is None else f", {union_offset} as __i__"
+    base += f" FROM {data_table} WHERE feature_view_name = {feature_view_name} AND entity_id = %s IN "
+    base += f"(SELECT version_id FROM {data_table} WHERE feature_view_name = {feature_view_name});"
+    return base
+
+
+def _execute_query_with_retry(
+    cur: Cursor,
+    conn: Connection,
+    query: str,
+    values: Union[List, Tuple],
+    retries: int = 3,
+    progress: Optional[Callable[[int], Any]] = None,
+    commit: bool = True,
+    propagate_exceptions: bool = False,
+    exponential: bool = False
+) -> bool:
+    for _ in range(retries):
+        try:
+            cur.execute(query, values)
+            if commit:
+                conn.commit()
+            if progress:
+                progress(1)
+            return True
+        except pymysql.Error as e:
+            if e.args[0] == MYSQL_DEADLOCK_ERR:
+                time.sleep(0.5 * (2 ** retries) if exponential else 0.5)
+            elif propagate_exceptions:
+                raise e
+            else:
+                conn.rollback()
+                logging.error("Error %d: %s" % (e.args[0], e.args[1]))
+                return False
+    return False
+
+
+def _get_feature_view_version_id(
+    cur: Cursor,
+    conn: Connection,
+    config: MySQLOnlineStoreConfig,
+    feature_view_name: str,
+) -> Optional[int]:
+    table, version_id = config.feature_version_table, None
+    cur.execute(
+        "SELECT version_id FROM %s WHERE feature_view_name = %s;",
+        (table, feature_view_name)
+    )
+    record = cur.fetchone()
+    if record is not None:
+        version_id = record[0]
+    else:
+        logging.info(f'FeatureView ({feature_view_name} not found in {table}).')
+    conn.commit()
+    return version_id
+
+
+def insert_feature_view_version(
+    cur: Cursor,
+    conn: Connection,
+    config: MySQLOnlineStoreConfig,
+    feature_view_name: str,
+    version_id: int,
+    lock_row: bool = False,
+    commit: bool = True
+) -> None:
+    write_query, table = "INSERT INTO %s VALUES (%s, %s, %s,)", config.feature_version_table
```

Comment:
will need to test this with non-ascii characters and strings to make sure behavior is as expected

Reply:
@franciscojavierarceo What do you mean?